be/src/util/jni-util.h

```diff
@@ -72,138 +59,1114 @@ class JniUtil {
-                                      jobject* hashmap_object);
-    static Status convert_to_cpp_map(JNIEnv* env, jobject map,
-                                     std::map<std::string, std::string>* resultMap);
-    static size_t get_max_jni_heap_memory_size();
-    static Status clean_udf_class_load_cache(const std::string& function_signature);
 
 private:
-    static void parse_max_heap_memory_size_from_jvm(JNIEnv* env);
-    static Status GetJNIEnvSlowPath(JNIEnv** env);
-    static Status init_jni_scanner_loader(JNIEnv* env);
+    // Thread-local cache of the JNIEnv for this thread.
+    static __thread JNIEnv* tls_env_;
 
-    static bool jvm_inited_;
-    static jclass internal_exc_cl_;
-    static jclass jni_native_method_exc_cl_;
+    //for exception
     static jclass jni_util_cl_;
     static jmethodID throwable_to_string_id_;
     static jmethodID throwable_to_stack_trace_id_;
-    static jmethodID get_jvm_metrics_id_;
-    static jmethodID get_jvm_threads_id_;
-    static jmethodID get_jmx_json_;
-    // JNI scanner loader
-    static jobject jni_scanner_loader_obj_;
-    static jmethodID jni_scanner_loader_method_;
-    // Thread-local cache of the JNIEnv for this thread.
-    static __thread JNIEnv* tls_env_;
-    static jlong max_jvm_heap_memory_size_;
-    static jmethodID _clean_udf_cache_method_id;
 };
 
-/// Helper class for lifetime management of chars from JNI, releasing JNI chars when
-/// destructed
-class JniUtfCharGuard {
+enum RefType { Local, Global };
+
+enum BufferType { Chars, ByteArray };
+
+template <RefType Ref>
+struct RefHelper {};
+
+template <>
+struct RefHelper<Local> {
+    static jobject create(JNIEnv* env, jobject obj) { return env->NewLocalRef(obj); }
+
+    static void destroy(JNIEnv* env, jobject obj) { env->DeleteLocalRef(obj); }
+
+    static Status get_env(JNIEnv** env) {
+        // Get the JNIEnv* corresponding to current thread.
+        return Jni::Env::Get(env);
+    }
+};
+
+template <>
+struct RefHelper<Global> {
+    static jobject create(JNIEnv* env, jobject obj) { return env->NewGlobalRef(obj); }
+
+    static void destroy(JNIEnv* env, jobject obj) { env->DeleteGlobalRef(obj); }
+
+    static Status get_env(JNIEnv** env) { return Jni::Env::Get(env); }
+};
+
+template <RefType Ref>
+class Object;
+
+template <RefType Ref>
+class Class;
+
+class MethodId {
+public:
+    MethodId() = default;
+    bool uninitialized() const { return _id == nullptr; }
+
+    template <RefType U>
+    friend class Object;
+
+    template <RefType U>
+    friend class Class;
+
+private:
+    jmethodID _id = nullptr;
+};
+
+class FieldId {
+public:
+    FieldId() = default;
+    bool uninitialized() const { return _id == nullptr; }
+
+    template <RefType U>
+    friend class Object;
+
+    template <RefType U>
+    friend class Class;
+
+private:
+    jfieldID _id = nullptr;
+};
+
+enum CallTag {
+    ObjectMethod,
+    IntMethod,
+    LongMethod,
+    VoidMethod,
+    BooleanMethod,
```

Comment:
jboolean z;
    jbyte    b;
    jchar    c;
    jshort   s;
    jint     i;
    jlong    j;
    jfloat   f;
    jdouble  d;
    jobject  l;

those all of types, maybe could add those type once